feat!: switch to canonical json for hashing (#683)
tkhq/qos: src/qos_json/SPEC.md

# QOS Canonical JSON

QOS uses JSON for protocol messages and signing payloads. Any JSON bytes that
are hashed, signed, or verified MUST be canonicalized using QOS-normalized RFC
8785-style canonical JSON. QOS intentionally differs from RFC 8785 by banning
all JSON number tokens.

## Canonicalization

- Implementations MUST parse JSON into a JSON value and re-encode that value as
  QOS-normalized RFC 8785-style canonical JSON before hashing or signing.
- Implementations MUST NOT hash or sign raw inbound JSON bytes. Equivalent JSON
  documents can differ in whitespace or field order.
- Implementations that serialize typed values for hashing or signing MUST use

>   the same canonical JSON encoder used for parsed JSON values.

Reviewer: ```suggestion
  the same canonical JSON encoding scheme used for parsed JSON values.
```
maybe? encoder implies a specific lib but we don't require a lib just a scheme

> - JSON used only for non-signed, non-hashed application messages MAY use normal

Reviewer: i think we can omit this bullet point

  JSON serialization.
- QOS normalization treats object members with `null` values as unset fields
  and MUST omit those object members before canonicalization. This means typed
  optional fields that serialize as `null` are omitted by the canonical encoder.
  Other `null` values are preserved. In particular, null array elements are not
  omitted because that would change array length and indexes.
- JSON object properties MUST be sorted recursively by raw, unescaped property
  names encoded as UTF-16 code units, as specified by RFC 8785 section 3.2.3.
  This matters for non-ASCII names: for example, the RFC 8785 sample order is
  carriage return, `1`, U+0080, `ö`, `€`, grinning-face emoji, and U+FB33.

> - JSON number tokens MUST NOT appear in signed, hashed, or verified JSON.

Reviewer: ```suggestion
- JSON number tokens MUST NOT appear in JSON canonical to this spec.
```

Reviewer: I don't feel strongly but I think all these MUST clauses apply to this this schema/spec for canonical json. Then the utility of this spec is for signed/hashed/verified json 
I could be wrong on how to normally write this style of doc

>   Implementations MUST terminate canonical serialization if a JSON number token

Reviewer: we should note if this requires the serde trait

Author: Not sure what you mean by this?

Reviewer: I just meant explicitly calling out: `#[serde(with = "qos_json::string_number")]`. Now that you have a test vector for it though I think its less of a concern

  is encountered.

## Signing Object Schemas

- Each signing object MUST have a specified JSON schema. This specification is
  not, by itself, a rule for choosing every field representation in every
  signing object.
- The schema for each signing object MUST specify object field names, enum
  representations, binary encodings, numeric encodings, and which fields are
  optional.
- Unset optional object fields are omitted by QOS canonicalization when they
  serialize as `null`. Rust schemas SHOULD use `#[serde(default)]` on optional
  fields that readers must accept when absent.
- A signing object schema MAY specify that empty map/object fields are omitted
  with type-specific serialization rules.
- New fields SHOULD be optional unless every deployed reader can tolerate the
  field. Append-friendly schemas rely on canonical null-member omission so older
  and newer writers produce the same payload until the new field is populated.

> - Binary values SHOULD be lowercase hex strings unless the signing object schema

Reviewer: maybe we can call out qos_hex::serde pattern

  explicitly specifies another representation.
- Enums SHOULD use serde's externally tagged representation unless the signing
  object schema explicitly specifies another representation.

> - Numeric domain values MUST be represented as decimal strings.

Reviewer: I don't really understand what this means. Will look for test vectors below

- Floating point values MUST NOT be introduced into QOS signing payloads.

> ## Test Vectors

Reviewer: wdyt of test vectors that show schema -> json -> hash?
basically the idea being have enough test vectors you can plug it into an LLM and have it produce a correct impl for another language. We could also link out to unit tests if we have some that would function as good test vectors
